feat(cloud-function): redirect non-canonical URLs

caugner opened the pull request:
## Summary

(MP-1111)

### Problem

We currently serve the same content independent of capitalization:

- https://developer.mozilla.org/en-US/docs/web (incorrect capitalization)
- https://developer.mozilla.org/en-US/docs/Web (correct capitalization)

### Solution

~~Fetch the `metadata.json` and match the URL (excluding search parameters) against the `mdn_url` in it.~~

Build a canonical map to determine the canonical URL for a requested URL, and redirect if the requested URL differs.

---

## How did you test this change?

1. Ran `npm ci && npm start` in `/cloud-function`.
2. Verified that `http://localhost:5100/en-US/docs/web?foo=bar` redirects to `http://localhost:5100/en-US/docs/Web?foo=bar`.

caugner committed:
feat(cloud-function): redirect capitalization

fiji-flo reviewed (commented):
Can we use a way that doesn't add another fetch please?

caugner committed:
feat(cloud-function): redirect capitalization (v2)

1. Build `sitemap.txt`, with all pages from all sitemaps.
2. Build `sitemap.json` for the Cloud Function, similar to the `redirects.json`.
3. Use middleware to redirect all URLs that match after normalization.

caugner committed:
Revert "feat(cloud-function): redirect capitalization"

This reverts commit 1c769894d773ef6e265638badc6207b9c845df17.

caugner committed:
Merge branch 'main' into MP-1111-redirect-capitalization

caugner commented:
> Can we use a way that doesn't add another fetch please?

Here's my second approach, utilizing a global sitemap.

caugner committed:
refactor(cloud-function): rename redirect{Capitalization => NonCanonicals}

caugner committed:
fix(cloud-function): parse url with URL() + support query strings

caugner committed:
fix(cloud-function): compare target with original source

caugner committed:
fixup! feat(cloud-function): redirect capitalization (v2)

fiji-flo commented on cloud-function/src/utils.ts:
```diff
@@ -59,3 +59,7 @@ const ANY_ATTACHMENT_REGEXP = createRegExpFromExtensions(
   return ANY_ATTACHMENT_REGEXP.test(url);
 }
+
+export function normalizeSlug(slug: string) {
+  return slug.toLowerCase().replace(/\/$/, "");
+}
```
This is also used in `buildSitemaps.ts` and it should be called normalize path or something.

caugner replied:
Good point and catch, fixed in 038740d330 and 237126d022.

caugner committed:
chore(cloud-function): reuse normalizeSlug()

caugner committed:
refactor(cloud-function): rename normalize{Slug => Path}()

caugner committed:
refactor(cloud-function): use object for sitemap

caugner commented on cloud-function/src/build-sitemap.ts:
```diff
@@ -0,0 +1,39 @@
+import { readFile, stat, writeFile } from "node:fs/promises";
```
> Generally, why not create the json instead of the .txt?

I went for text format, because [text sitemaps](https://developers.google.com/search/docs/crawling-indexing/sitemaps/build-sitemap#text) are a thing and it's a simpler format with fewer assumptions than a JSON, although I'm not technically adhering to the sitemap convention of using fully qualified URLs.

> I mainly want be aware of what's the contract between building and the cloud function and I think the json makes a better api.

Are you suggesting to publish a `sitemap.json` with a mapping of normalized path to canonical path, to be used by other consumers potentially? That would actually be my concern that other parties start using this JSON if they come across it. On the other hand, would we change it?

> If there's a strong reason to keep this I'd just have ask to use a Object instead of the Map since we're putting it into and Object any way.

Done in d2e58ca1d4.

fiji-flo replied:
Yes I was suggesting to build the same `sitemap.json` as in this file. I wouldn't call it sitemap though.

I would not _publish_ that file. Is there a benefit for having the sitemap as text and as XML? If we go for only txt sitemap I'm all in for this. Otherwise this ads complexity and we potentially have to support two sitemaps formats from now on.

caugner replied:
I don't mind renaming, and I could surely build that map as part of the build, but I don't think it reduces complexity:

1. I need to get all URLs from all sitemaps, so either I gather them as a side-effect of building the sitemaps, or I go over all (gzip-compressed) XML sitemaps later, parsing out the URLs.
2. We build and deploy the Cloud Function after we sync the build to Cloud Storage, and `gsutil` doesn't support exclusion very well (cf. the `plain.html` files that we sync), so we might need to prevent access to that file in the Cloud Function.
3. We definitely want to keep using the XML sitemap for search engines (as advised in `robots.txt`), because the text format doesn't support modification dates.
4. We would need to extract the `normalizePath()` function into a lib shared by build and cloud function, and I assume this would no longer be possible with rari. We could then duplicate the logic, but it's still risky.

Overall, I would recommend to keep the current approach.

caugner committed:
refactor(cloud-function): rename {sitemap => canonicals}

caugner committed:
refactor(build): extract makeSitemapIndexTXT

caugner committed:
fix(build): distinguish sitemap.{txt,xml.gz}

caugner committed:
refactor(build): extract buildSitemapIndex{TXT,XML}

caugner committed:
Merge branch 'main' into MP-1111-redirect-capitalization

caugner committed:
Merge branch 'main' into MP-1111-redirect-capitalization

caugner committed:
Merge branch 'main' into MP-1111-redirect-capitalization

fiji-flo commented on cloud-function/src/app.ts:
```diff
@@ -39,6 +40,7 @@ router.all(
 router.all("/submit/mdn-yari/*", requireOrigin(Origin.main), proxyTelemetry);
 router.all("/pong/*", requireOrigin(Origin.main), express.json(), proxyPong);
 router.all("/pimg/*", requireOrigin(Origin.main), proxyPong);
+router.use(redirectNonCanonicals);
```
Can we reorder the routes here? Move everything that doesn't need the `redirectNonCanonicals` up (I guess only get / and the assests).

Otherwise ready to go.

caugner replied:
@fiji-flo I moved `redirectNonCanonicals` down instead (and the `search-index.json` route up next to the assets route), to make the change less invasive: https://github.com/mdn/yari/pull/11151/files/55d08d37374c549d154529d8228dabcca4fc9a2d..45501138daec268e2fdb0dfa7a6ab9229f99f4b3

caugner committed:
Merge branch 'main' into MP-1111-redirect-capitalization

caugner committed:
chore(build): remove unused fdir import

caugner committed:
chore(cloud-function): move canonical redirect down + search-index route up

caugner committed:
chore(cloud-function): add comments

caugner commented:
@fiji-flo See https://github.com/mdn/yari/pull/11151#discussion_r1648782304.

fiji-flo reviewed (approved):
👍
